FreeRTOS-Plus/Source/FreeRTOS-IoT-Libraries-LTS-Beta2/c_sdk/platform/freertos/transport/src/tls_freertos.c

```diff
@@ -0,0 +1,611 @@
+        ( tlsStatus != ( BaseType_t ) MBEDTLS_ERR_SSL_WANT_WRITE ) )
+    {
+        if( tlsStatus == 0 )
+        {
+            LogInfo( ( "(Network connection %p) TLS close-notify sent.",
+                       pNetworkContext ) );
+        }
+        else
+        {
+            LogError( ( "(Network connection %p) Failed to send TLS close-notify: mbedTLSError= %s : %s.",
+                        pNetworkContext,
+                        mbedtlsHighLevelCodeOrDefault( tlsStatus ),
+                        mbedtlsLowLevelCodeOrDefault( tlsStatus ) ) );
+        }
+    }
+    else
+    {
+        /* WANT_READ and WANT_WRITE can be ignored. Logging for debugging purposes. */
+        LogInfo( ( "(Network connection %p) TLS close-notify sent; ",
+                   "received %s as the TLS status can be ignored for close-notify."
+                   ( tlsStatus == MBEDTLS_ERR_SSL_WANT_READ ) ? "WANT_READ" : "WANT_WRITE",
+                   pNetworkContext ) );
+    }
+
+    /* Call socket shutdown function to close connection. */
+    Sockets_Disconnect( pNetworkContext->tcpSocket );
+
+    /* Free mbed TLS contexts. */
+    sslContextFree( &( pNetworkContext->sslContext ) );
+
+    /* Free the contexts for random number generation. */
+    mbedtls_ctr_drbg_free( &ctrDrgbContext );
+    mbedtls_entropy_free( &entropyContext );
+
+    /* Clear the mutex functions for mbed TLS thread safety. */
+    mbedtls_threading_free_alt();
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t TLS_FreeRTOS_recv( NetworkContext_t * pNetworkContext,
+                           void * pBuffer,
+                           size_t bytesToRecv )
+{
+    int32_t tlsStatus = 0;
+
+    tlsStatus = ( int32_t ) mbedtls_ssl_read( &( pNetworkContext->sslContext.context ),
+                                              pBuffer,
+                                              bytesToRecv );
+
+    if( ( tlsStatus == MBEDTLS_ERR_SSL_TIMEOUT ) ||
+        ( tlsStatus == MBEDTLS_ERR_SSL_WANT_READ ) ||
+        ( tlsStatus == MBEDTLS_ERR_SSL_WANT_WRITE ) )
+    {
+        /* Mark these set of errors as a timeout for the libraries to retry
+         * the read. */
+        tlsStatus = 0;
+    }
+    else if( tlsStatus < 0 )
+    {
+        LogError( ( "Failed to read data: errorStatus=%d", tlsStatus ) );
+    }
+    else
+    {
+        /* Empty else marker. */
+    }
+
+    return tlsStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t TLS_FreeRTOS_send( NetworkContext_t * pNetworkContext,
+                           const void * pBuffer,
+                           size_t bytesToSend )
+{
+    int32_t tlsStatus = 0;
+
+    tlsStatus = ( int32_t ) mbedtls_ssl_write( &( pNetworkContext->sslContext.context ),
+                                               pBuffer,
+                                               bytesToSend );
+
+    if( ( tlsStatus == MBEDTLS_ERR_SSL_TIMEOUT ) ||
```

Comment:
We have 3 different errors here but we loose which one happened, I think we need a log_error here that will tell you if it was a timeout, read or write error.

Reply:
These errors are marked as a timeout so that the libraries can retry a recv/send, if needed. As per the transport interface definition, We will have 3 types of returns for a send of receive.  
Pasting the documentation for TLS_FreeRTOS_recv from tls_freertos.h below.
@return Number of bytes (> 0) received if successful;
 * 0 if the socket times out without reading any bytes;
 * negative value on error.

The decision of whether to retry or not for timeout is completely upto the library to decide. It may not retry if the timeout for the API(passed as a parameter for the API) has exceeded while waiting for a send or receive to complete. 

I will update the comments to reflect this.